### ISSUE TITLE
Change in DNS resolution behaviour in 25.0

### ISSUE THREAD

**Issue by ap-wtioit** [moby/moby#47333]
### Description

An internal network created in 24.0 did not resolve outside domains while in 25.0 DNS queries to the outside are no longer blocked. This allows DNS Data Exfiltration in networks that where safe in 24.0.

I think this should be at least stated in the Changelog for 25.0.

### Reproduce

1. `docker network create --internal test_network_internal`
2. `docker run --rm --network test_network_internal curlimages/curl --connect-timeout 2 -v 'http://www.google.com/'`

### Expected behavior

in Docker version 24.0.2, build cb74dfc, DNS queries would not be made to the outside
```
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0* Resolving timed out after 2001 milliseconds
  0     0    0     0    0     0      0      0 --:--:--  0:00:02 --:--:--     0
* Closing connection
curl: (28) Resolving timed out after 2001 milliseconds
```
while in Docker version 25.0.2, build 29cf629, the DNS query is made to the outside
```
% Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* Host [www.google.com:80](http://www.google.com/) was resolved.
* IPv6: 2a00:1450:400d:80c::2004
* IPv4: [142.250.180.196](http://142.250.180.196/)
*   Trying [2a00:1450:400d:80c::2004]:80...
* Immediate connect fail for 2a00:1450:400d:80c::2004: Address not available
*   Trying 142.250.180.196:80...
* Immediate connect fail for 142.250.180.196: Network unreachable
* Failed to connect to [www.google.com](http://www.google.com/) port 80 after 6 ms: Couldn't connect to server
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
* Closing connection
curl: (7) Failed to connect to [www.google.com](http://www.google.com/) port 80 after 6 ms: Couldn't connect to server
```

### docker version

```bash
Client: Docker Engine - Community
 Version:           25.0.2
 API version:       1.44
 Go version:        go1.21.6
 Git commit:        29cf629
 Built:             Thu Feb  1 00:23:02 2024
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          25.0.2
  API version:      1.44 (minimum version 1.24)
  Go version:       go1.21.6
  Git commit:       fce6e0c
  Built:            Thu Feb  1 00:23:02 2024
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.28
  GitCommit:        ae07eda36dd25f8a1b98dfbf587313b99c0190bb
 runc:
  Version:          1.1.12
  GitCommit:        v1.1.12-0-g51d5e94
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
Client: Docker Engine - Community
 Version:    25.0.2
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.12.1
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.24.5
    Path:     /usr/libexec/docker/cli-plugins/docker-compose

Server:
 Containers: 90
  Running: 57
  Paused: 0
  Stopped: 33
 Images: 215
 Server Version: 25.0.2
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: ae07eda36dd25f8a1b98dfbf587313b99c0190bb
 runc version: v1.1.12-0-g51d5e94
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 6.5.0-15-generic
 Operating System: Ubuntu 23.10
 OSType: linux
 Architecture: x86_64
 CPUs: 16
 Total Memory: 61.94GiB
 Name: redacted
 ID: WFOT:GN2I:KKU2:AHXI:QIJS:JHXO:5CGH:UYXZ:MADD:JK4W:K7Q3:ZAE7
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: true
 Default Address Pools:
   Base: 172.16.0.0/12, Size: 26
```

**Comment by ap-wtioit** [comment]
I checked with security@docker.com they are not considering this a security issue and consider this expected behaviour.

If it is expected a change in behaviour this big should be worth a changelog entry.

**Comment by ap-wtioit** [comment]
Just tested an on our servers this is independent of the contents of /etc/resolv.conf (`nameserver 127.0.0.53` vs `nameserver 8.8.8.8`) when using this version:
```
Client: Docker Engine - Community
 Version:           24.0.5
 API version:       1.43
 Go version:        go1.20.6
 Git commit:        ced0996
 Built:             Fri Jul 21 20:35:35 2023
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          24.0.5
  API version:      1.43 (minimum version 1.12)
  Go version:       go1.20.6
  Git commit:       a61e2b4
  Built:            Fri Jul 21 20:35:35 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.22
  GitCommit:        8165feabfdfe38c65b599c4993d227328c231fca
 runc:
  Version:          1.1.8
  GitCommit:        v1.1.8-0-g82f18fe
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```
and the having the following info
```
Client: Docker Engine - Community
 Version:    24.0.5
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.11.2
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.20.2
    Path:     /usr/libexec/docker/cli-plugins/docker-compose

Server:
 Containers: 6
  Running: 6
  Paused: 0
  Stopped: 0
 Images: 94
 Server Version: 24.0.5
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 8165feabfdfe38c65b599c4993d227328c231fca
 runc version: v1.1.8-0-g82f18fe
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 6.1.0-9-amd64
 Operating System: Debian GNU/Linux 12 (bookworm)
 OSType: linux
 Architecture: x86_64
 CPUs: 16
 Total Memory: 61.95GiB
 Name: redacted
 ID: 38091a89-03d8-4cce-90e2-a3fbcb07137b
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: true
 Default Address Pools:
   Base: 172.16.0.0/12, Size: 26
   Base: 192.168.0.0/16, Size: 26
```
we get the timeout for resolving.

whereas the server with docker 25.0 always (with namerserver 127.0.0.53 vs nameserver 8.8.8.8) has the connection refused after resolving.

**Comment by akerouanton** [comment]
> whereas the server with docker 25.0 always (with namerserver 127.0.0.53 vs nameserver 8.8.8.8) has the connection refused after resolving.

Did you restart the engine between both tests? Docker doesn't detect changes made to `/etc/resolv.conf` and thus need a full restart to take them into account.

Can you paste the output of `ip route show` and `cat /etc/resolv.conf` from within a container connected to an internal network (ideally for both of your v24 and v25 environments)?

**Comment by ap-wtioit** [comment]
> Did you restart the engine between both tests?

No i did not. I knew something was missing. Sorry.

I finally made a small [ansible / vagrant test project](https://github.com/ap-wtioit/docker_network_test) that spawns up to ubuntu mantic VMs and installs docker 24 and 25. And I can confirm in those clean environments the only change to the internal networks is the missing gateway that's mentioned in the changelog (`ip route show` shows that). That leads to way faster failures on outgoing connections in docker 25 in comparison to docker 24. And as you said only if there is a non-loopback DNS used (e.g. 8.8.8.8) then docker has this behaviour of blocking/not routing DNS traffic from internal networks in both versions 24 and 25.

And i guess our servers with docker 25 only experience this "changed" behaviour because I updated them recently and restarted them so systemd-resolved has set a loopback DNS in /etc/resolv.conf which i didn't catch.

Sorry again for bothering you.
